### PR TITLE
[#924] Add ID to Dates

### LINKED ISSUE #924: [Export] Make dates have id
The `dates` field format should be as documented here https://inveniordm.docs.cern.ch/reference/metadata/#dates-0-n (no need to pass the `title` subfield though*). So something like:

```json
{
  "dates": [{
    "date": "1939/1945",
    "type": {
      "id": "other",
    },
    "description": "A date"  // optional
  }],
}
```

The `id` must be taken from: https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/fixtures/data/vocabularies/date_types.yaml

`*`: This is admittedly not good documentation. The Metadata Reference section is really meant for librarians to understand what is stored in their instance. The REST API has sligthly different input / output details that are not captured there.

### PR DESCRIPTION
**Pull request by sharpattack**
For 'type'. closes #924

### CODE REVIEW

File: app/models/invenio_rdm_record_converter.rb [hunk 1]
```diff
@@ -180,7 +180,7 @@ def invenio_metadata
       "publication_date": format_publication_date(@generic_file.date_created.shift || @generic_file.date_uploaded.to_s),
       "subjects": SUBJECT_SCHEMES.map{ |subject_type| subjects_for_scheme(@generic_file.send(subject_type), subject_type) }.compact.flatten,
       "contributors": contributors(@generic_file.contributor),
-      "dates": @generic_file.date_created.map{ |date| {"date": date, "type": "other", "description": "When the item was originally created."} },
+      "dates": @generic_file.date_created.map{ |date| {"date": date, "type": {"id": "other"}, "description": "When the item was originally created."} },
```
Comment (fenekku) [inline app/models/invenio_rdm_record_converter.rb:183]: Oh this is always the date of creation? Then, we should use this id: https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/fixtures/data/vocabularies/date_types.yaml#L21

File: spec/models/invenio_rdm_record_converter_spec.rb [hunk 2]
```diff
@@ -104,7 +104,7 @@
             },
             "role": {"id": InvenioRdmRecordConverter::ROLE_OTHER}
           }],
-          "dates": [{"date": "2021-1-1", "type": "other", "description": "When the item was originally created."}],
+          "dates": [{"date": "2021-1-1", "type": {"id":"other"}, "description": "When the item was originally created."}],
```
Comment (fenekku) [inline spec/models/invenio_rdm_record_converter_spec.rb:107]: The date format should follow EDTF level 0 https://inveniordm.docs.cern.ch/reference/metadata/#dates-0-n which means double digit resolution i.e. it should be `"2021-01-01"`.

**Review by fenekku — COMMENTED** [review]
More tweaks :paintbrush:

**Comment by Meowcenary** [comment]
Seems like @fenekku has covered all the necessary changes. Looks pretty straightforward to me so approving pending changes.

**Review by Meowcenary — APPROVED** [review]

File: spec/models/invenio_rdm_record_converter_spec.rb (hunk 2 above)
Comment (sharpattack) [reply spec/models/invenio_rdm_record_converter_spec.rb:107]: Will the import still allow any dates? Or is it strictly EDTF level 0? This a free text field in DigitalHub, unfortunately.

Comment (fenekku) [reply spec/models/invenio_rdm_record_converter_spec.rb:107]: No, the import only allows EDTF level 0 dates.

**Comment by sharpattack** [comment]
@saragon02 There are dates formatted as "mm/dd/yy". Because we need the format to be in EDTF level 0, the year has to be "YYYY", fully 'spelled' out. Is it okay if we draw the line at... 22? So "yy" in the range of 00-22 will be in the 2000s, and the range of 23-99 will be in the 1900s? A closer look at the recent metadata dump sent to Liping may help answer this.

**Comment by saragon02** [comment]
@sharpattack If the purpose is for automating conversion of Date Created dates to YYYY format, I think that may not work because I'm pretty confident we'll have some pre-1900 creation dates for some items (Civil War era things, certain photos & letters, etc.)

**Comment by sharpattack** [comment]
Thanks to @saragon02 for the help! +1 to using OpenRefine. I was using excel to examine a recent metadata dump, and excel decides to reformat dates to their liking. There are no dates in formatted with two digits, as "yy". 
@fenekku and @Meowcenary the changes are now ready for review.

File: app/models/invenio_rdm_record_converter.rb [hunk 3]
```diff
@@ -440,6 +445,19 @@ def normalize_date(date_string)
     end
   end
 
+  def rearrange_year(date_array)
+    if date_array[2].length == 2
+      # drawing the line at 22 as the year for when we no longer refer to 1900s
+      if date_array[2].to_i > 22
```
Comment (Meowcenary) [inline app/models/invenio_rdm_record_converter.rb]: I want to put the 22 into a constant or something similar to describe what it's doing, but I honestly, I can't really think of a good name. The comment is very helpful though.

File: app/models/invenio_rdm_record_converter.rb [hunk 4]
```diff
@@ -375,6 +375,11 @@ def format_publication_date(publication_date)
 
   def normalize_date(date_string)
     split_date = date_string.split(/[-,\/ ]/).map(&:downcase)
+    # date format starts with month first
+    if (!split_date.blank? && split_date[0].length < 3)
```
Comment (Meowcenary) [inline app/models/invenio_rdm_record_converter.rb:379]: Thanks for figuring this out! I was having a really hard time with these kinds of dates previously.

**Review by Meowcenary — APPROVED** [review]
Looks good to me Austin. Thanks for working on this.

**Review by fenekku — APPROVED** [review]
So the case of a year with 2 digits never really happened?

File: app/models/invenio_rdm_record_converter.rb (hunk 3 above)
Comment (sharpattack) [reply app/models/invenio_rdm_record_converter.rb]: Haha, well comment or constant, we don't need this anymore. There are no dates with just "yy". They all include the century, "YYYY".

File: app/models/invenio_rdm_record_converter.rb (hunk 4 above)
Comment (sharpattack) [reply app/models/invenio_rdm_record_converter.rb:379]: Thanks to you for getting this started! I only added a small piece, while you did the heavy lifting with the dates.

**Comment by sharpattack** [comment]
*sigh* There were no years with 2 digits. Thankfully, though.

**Comment by fenekku** [comment]
ok

**Pull request merged**